### PR TITLE
Merge in agri-vis challenge stuff (HDF5, autoencoders, multi-label classification, ...)

### PR DESCRIPTION
**Pull request by plstcharles**
Lots of medium-sized-changes, should hopefully not break too much of the old stuff, but needed to move towards more unsupervised learning stuff with bigearthnet, AEs, and self-supervised tasks.

@fmigneault feel free to ask for changes if you see something weird --- tests pass, but it's probably not totally clean. Once this is in, we should probably bump the version...

### CODE REVIEW

File: cluster_orion_hunt.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
```
Comment (fmigneault) [inline cluster_orion_hunt.sh]: Probably should move all those cluster scripts into a ./cluster directory to kept root clean.

File: conda-env.yml [hunk 2]
```diff
@@ -6,8 +6,13 @@ dependencies:
   - cython
   - ddt
   - gitdb
+  - gitpython
+  - h5py
+  - hdf5plugin
+  - lz4
+  - matplotlib
+  - numpy
```
Comment (fmigneault) [inline conda-env.yml:14]: doesn't some of these cause conflicts with packages already in requirements.txt?

File: thelper/draw.py [hunk 3]
```diff
@@ -703,9 +718,9 @@ def draw_bbox(image, tl, br, text, color, box_thickness=2, font_thickness=1,
     return win_name, image
 
 
-def draw_bboxes(images,                 # type: thelper.typedefs.ImageArray
-                preds=None,             # type: Optional[thelper.typedefs.BoundingBoxArray]
-                bboxes=None,            # type: Optional[thelper.typedefs.BoundingBoxArray]
+def draw_bboxes(images,
+                preds=None,
+                bboxes=None,
                 color_map=None,         # type: Optional[thelper.typedefs.ClassColorMap]
```
Comment (fmigneault) [inline thelper/draw.py:724]: Why remove the typings?

File: thelper/ifaces.py [hunk 4]
```diff
@@ -77,18 +82,22 @@ class ClassNamesHandler(ABC):
...
 
     @property
-    def class_names(self):
+    def class_names(self) -> typing.List[str]:
         """Returns the list of class names considered "of interest" by the derived class."""
         return self._class_names
 
     @class_names.setter
-    def class_names(self, class_names):
+    def class_names(self, class_names) -> None:
```
Comment (fmigneault) [inline thelper/ifaces.py]: `class_names: typing.Dict[str, typing.Union[int,str]]` ?

File: thelper/ifaces.py [hunk 5]
```diff
@@ -162,19 +176,27 @@ class FormatHandler(ABC):
...
+    ) -> None:
         self.format = None
         self.ext = None
         self.solve_format(format)
 
-    def solve_format(self, format):
-        # type: (Optional[AnyStr]) -> None
+    def solve_format(
+        self,
+        format: typing.Optional[typing.AnyStr]
```
Comment (fmigneault) [inline thelper/ifaces.py]: optional should have format=None

File: thelper/nn/resnet.py [hunk 6]
```diff
@@ -27,66 +27,83 @@ def _make_conv2d(self, *args, **kwargs):
...
+                 radius_channel=True, activation="relu"):
         super().__init__(inplanes, planes, stride, downsample, coordconv, radius_channel)
         self.conv1 = self._make_conv2d(inplanes, planes, kernel_size=3, stride=stride, padding=1, bias=False)
         self.bn1 = torch.nn.BatchNorm2d(planes)
-        self.relu = torch.nn.ReLU(inplace=True)
+        assert activation in ["relu", "leaky_relu"]
+        if activation == "relu":
+            self.activ = torch.nn.ReLU(inplace=True)
+        elif activation == "leaky_relu":
+            self.activ = torch.nn.LeakyReLU(inplace=True)
```
Comment (fmigneault) [inline thelper/nn/resnet.py]: I would move all duplicate activation code into another base class and add that base to all applicable modules.

File: thelper/nn/resnet.py [hunk 7]
```diff
@@ -262,7 +312,7 @@ def set_task(self, task):
             if self.fc is None or self.fc[len(self.fc) - 1].out_channels != num_classes:
                 if self.head_type == "fcn":
                     self.fc = torchvision.models.segmentation.fcn.FCNHead(self.out_features, num_classes)
-                else:
+                elif self.head_type == "deeplabv3":
                     self.fc = torchvision.models.segmentation.deeplabv3.DeepLabHead(self.out_features, num_classes)
```
Comment (fmigneault) [inline thelper/nn/resnet.py:306]: should we have an assert if None of the two?

File: thelper/nn/resnet.py [hunk 8]
```diff
@@ -365,3 +406,140 @@ def set_task(self, task):
                 torch.nn.Parameter(self.fc.weight.view(self.fc.out_features, self.out_features, 1, 1))
             self.finallayer.bias = torch.nn.Parameter(self.fc.bias)
         self.task = task
+
+
+class AEResNet(ResNet):
```
Comment (fmigneault) [inline thelper/nn/resnet.py]: I would write Autoencoder explicitly so it's easier to read in the config.

File: thelper/nn/resnet.py [hunk 9]
```diff
@@ -365,3 +406,140 @@ def set_task(self, task):
...
+        reconstruction = self.decoder_bottom(featmap)
+        return class_logits, reconstruction
+
+
+class FakeModule(torch.nn.Module):
+    def forward(self, inputs):
+        return inputs
+
+
+class AESkipResNet(ResNet):
```
Comment (fmigneault) [inline thelper/nn/resnet.py]: idem full name + UNet in name?

File: thelper/nn/utils.py [hunk 10]
```diff
@@ -182,6 +182,12 @@ def create_model(config, task, save_dir=None, ckptdata=None):
         model.config = model_params
     if hasattr(model, "summary"):
         model.summary()
+    param_count = get_learnable_param_count(model)
+    if param_count > 10 ** 5:
+        param_count_str = f"{(param_count // 10 ** 5) / 10}M"
+    else:
+        param_count_str = f"{param_count}"
```
Comment (fmigneault) [inline thelper/nn/utils.py:189]: when in range 100K>params>1M, this looks a bit weird, and most precision is lost
eg: 
```
f"{(125000 // 10 ** 5) / 10}M"
'0.1M'
```

File: thelper/session/base.py [hunk 11]
```diff
@@ -115,8 +115,9 @@ def __init__(self,
         self.skip_eval_iter = thelper.utils.get_key_def("skip_eval_iter", trainer_config, 0)
 
         # parse and prepare tbx stuff
-        self.use_tbx = thelper.utils.str2bool(thelper.utils.get_key_def(["use_tbx", "tbx", "use_tb", "tb", "tensorboard"],
-                                                                        trainer_config, False))
+        self.use_tbx = \
+            thelper.utils.get_key_def(["use_tbx", "tbx", "use_tb", "tb", "tensorboard"], trainer_config, False)
```
Comment (fmigneault) [inline thelper/session/base.py]: use variable for combinations instead of \

File: thelper/session/base.py [hunk 12]
```diff
@@ -211,8 +216,8 @@ def __init__(self,
             # parse user (custom) callback
             user_callback_keys = [f"{cname}_iter_callback", f"{cname}_callback", "callback"]
             user_callback = \
-                thelper.utils.get_key_def(user_callback_keys,
-                                          trainer_config)  # type: Optional[thelper.typedefs.IterCallbackType]
+                thelper.utils.get_key_def(user_callback_keys, trainer_config) \
+                # type: Optional[thelper.typedefs.IterCallbackType]
```
Comment (fmigneault) [inline thelper/session/base.py]: better like this?
```
user_callback = thelper.utils.get_key_def(
    user_callback_keys, trainer_config
)  # type: Optional[thelper.typedefs.IterCallbackType]
```

File: thelper/tasks/utils.py [hunk 13]
```diff
@@ -4,19 +4,17 @@
...
 
 import thelper.typedefs
 import thelper.utils
 
 logger = logging.getLogger(__name__)
 
 
-def create_task(config):
-    # type: (Union[Task, thelper.typedefs.ConfigDict, str]) -> Task
+def create_task(config: typing.Union[thelper.typedefs.ConfigDict, typing.AnyStr]) -> "Task":
```
Comment (fmigneault) [inline thelper/tasks/utils.py:17]: It doesn't need `thelper.task.Task` ?

**Review by fmigneault — COMMENTED** [review]
Pretty good PR. I see only small fixes, the rest seem to remain back-compatible.

I would make 2 bumps, 
0.5.0 since it adds inference mode (rc2 feels ready, I've used it a few times and is stable for existing processes)

Then 0.6.0 for these PR changes

**Comment by fmigneault** [comment]
Other thing, probably need some update to docs about new models / tasks, etc.
There are a lot of `@@@@` todo still, makes it harder to introduce the package to new users.

**Comment by plstcharles** [comment]
true, all very good points --- i'll do some polish on those changes early next week (fingers crossed!)

**Comment by fmigneault** [comment]
We are using the repo actively at the moment, so all updates will be very helpful. Thanks!

File: conda-env.yml (hunk 2 above)
Comment (plstcharles) [reply conda-env.yml:14]: the overlap allows for a pip-only-installation, it's not super clean but it doesn't cause conflicts that would crash anything

File: thelper/draw.py (hunk 3 above)
Comment (plstcharles) [reply thelper/draw.py:724]: no clue, types were bad maybe, added back in with correct ones

File: thelper/nn/resnet.py (hunk 6 above)
Comment (plstcharles) [reply thelper/nn/resnet.py]: solved with a module function for now, this whole file is just hacks that should not exist

File: thelper/nn/resnet.py (hunk 7 above)
Comment (plstcharles) [reply thelper/nn/resnet.py:306]: not necessarily --- a parent class might call that function, leave it default out to none, and implement its own layer after the call (pretty hacky, but on par for this messy file)

File: thelper/nn/utils.py (hunk 10 above)
Comment (plstcharles) [reply thelper/nn/utils.py:189]: ahhh the good old days of small models are far...

File: thelper/tasks/utils.py (hunk 13 above)
Comment (plstcharles) [reply thelper/tasks/utils.py:17]: since it's defined in the same file, it works without the full spec (magicmagicmagic)

Comment (plstcharles) [reply thelper/tasks/utils.py:17]: (now in quotes because it's defined below the 1st use of the type)

**Comment by plstcharles** [comment]
cleanups are in, no bump yet however

File: conda-env.yml (hunk 2 above)
Comment (fmigneault) [reply conda-env.yml:14]: Ok, I was wondering because I have add some issues before where same package defined in requirements.txt and within docs/requirements.txt which also included `-r requirements.txt` line caused double entry conflict.
If conda handles it correctly, that's fine.

File: thelper/tasks/utils.py (hunk 13 above)
Comment (fmigneault) [reply thelper/tasks/utils.py:17]: 👍

**Review by fmigneault — APPROVED** [review]
All good. 
Only thing left is the changelog of all this and bump/tag.

**Comment by plstcharles** [comment]
change notes are in! can't bump on the brunch however, it adds a "-dev" postfix to the tag, so we'll have to merge first & bump on master

**Comment by fmigneault** [comment]
> 
> 
> change notes are in! can't bump on the brunch however, it adds a "-dev" postfix to the tag, so we'll have to merge first & bump on master

see PR #19
